liblwgeom/lwgeom.c

```diff
@@ -1606,55 +1629,46 @@ lwgeom_remove_repeated_points_in_place(LWGEOM *geom, double tolerance)
 		}
 		case MULTIPOINTTYPE:
 		{
-			double tolsq = tolerance*tolerance;
-			uint32_t i, j, n = 0;
-			LWMPOINT *mpt = (LWMPOINT *)(geom);
-			LWPOINT **out;
-			LWPOINT *out_stack[out_stack_size];
-			int use_heap = (mpt->ngeoms > out_stack_size);
-
-			/* No-op on empty */
-			if (mpt->ngeoms < 2)
-				return geometry_modified;
-
-			/* We cannot write directly back to the multipoint */
-			/* geoms array because we're reading out of it still */
-			/* so we use a side array */
-			if (use_heap)
-				out = lwalloc(sizeof(LWMPOINT *) * mpt->ngeoms);
-			else
-				out = out_stack;
+			double tolsq = tolerance * tolerance;
+			LWMPOINT *mpt = (LWMPOINT *)geom;
 
-			/* Inefficient O(n^2) implementation */
-			for (i = 0; i < mpt->ngeoms; i++)
+			for (uint32_t dim = 0; dim < 2; dim++)
 			{
-				int seen = 0;
-				LWPOINT *p1 = mpt->geoms[i];
-				const POINT2D *pt1 = getPoint2d_cp(p1->point, 0);
-				for (j = 0; j < n; j++)
+				qsort(mpt->geoms, mpt->ngeoms, sizeof(LWPOINT *), dim == 0 ? cmp_point_x : cmp_point_y);
+				for (uint32_t i = 0; i < mpt->ngeoms; i++)
 				{
-					LWPOINT *p2 = out[j];
-					const POINT2D *pt2 = getPoint2d_cp(p2->point, 0);
-					if (distance2d_sqr_pt_pt(pt1, pt2) <= tolsq)
+					if (mpt->geoms[i] == NULL)
+						continue;
+
+					const POINT2D *pti = getPoint2d_cp(mpt->geoms[i]->point, 0);
+					for (uint32_t j = i + 1; j < mpt->ngeoms; j++)
 					{
-						seen = 1;
-						break;
+						if (mpt->geoms[j] == NULL)
+							continue;
+
+						const POINT2D *ptj = getPoint2d_cp(mpt->geoms[j]->point, 0);
+						if ((dim == 0 ? ptj->x - pti->x : ptj->y - pti->y) > tolerance)
```

Comment:
Sounds like this should be in the outer loop condition and as a whole it should be a while loop - no need to init j if all points are farther away than dedup threshold

Reply:
hm, I don't think I can move this in outer loop since array is sparse and not null item in points array need to be found first